OpenRA.Game/WorldUtils.cs

```diff
@@ -41,6 +41,94 @@ public static IEnumerable<Actor> FindActorsInCircle(this World world, WPos origi
+
+			var actorsInSquare = world.ActorMap.ActorsInBox(finalTarget, finalSource);
+			var intersectedActors = new List<Actor>();
+
+			foreach (var currActor in actorsInSquare)
+			{
+				long actorWidth;
+				var healthInfo = currActor.Info.TraitInfoOrDefault<HealthInfo>();
+				if (healthInfo != null)
+					actorWidth = healthInfo.Radius.Length;
+				else
+					actorWidth = 0;
+
+				var projection = MinimumPointLineProjection(lineStart, lineEnd, currActor.CenterPosition);
+				var distance = (currActor.CenterPosition - projection).HorizontalLength;
+				var maxReach = (long)(actorWidth + lineWidth.Length);
+
+				if (distance <= maxReach)
+					intersectedActors.Add(currActor);
+			}
+
+			return intersectedActors;
+		}
+
+		/// <summary>
+		/// Find the point (D) on a line (A-B) that is closest to the target point (C).
+		/// </summary>
+		/// <param name="lineStart">The source point (tail) of the line</param>
+		/// <param name="lineEnd">The target point (head) of the line</param>
+		/// <param name="point">The target point that the minimum distance should be found to</param>
+		/// <returns>The WPos that is the point on the line that is closest to the target point</returns>
+		public static WPos MinimumPointLineProjection(WPos lineStart, WPos lineEnd, WPos point)
+		{
+			var squaredLength = (lineEnd - lineStart).HorizontalLengthSquared;
+
+			// Line has zero length, so just use the lineEnd position as the closest position.
+			if (squaredLength == 0)
+				return lineEnd;
+
+			// Consider the line extending the segment, parameterized as target + t (source - target).
+			// We find projection of point onto the line.
+			// It falls where t = [(point - target) . (source - target)] / |source - target|^2
+			// The normal DotProduct math would be (xDiff + yDiff) / dist, where dist = (target - source).LengthSquared;
+			// But in order to avoid floating points, we do not divide here, but rather work with the large numbers as far as possible.
+			// We then later divide by dist, only AFTER we have multiplied by the dotproduct.
+			var xDiff = ((long)point.X - (long)lineEnd.X) * ((long)lineStart.X - (long)lineEnd.X);
+			var yDiff = ((long)point.Y - (long)lineEnd.Y) * ((long)lineStart.Y - (long)lineEnd.Y);
+			long t = xDiff + yDiff;
+
+			// Beyond the 'target' end of the segment
+			if (t < 0)
+				return lineEnd;
+
+			// Beyond the 'source end of the segment
+			if (t > squaredLength)
+				return lineStart;
+
+			// Projection falls on the segment
+			// Using floating points, this would be: var newX = (int)((double)target.X + t * (double)(source.X - target.X));
+			// where t would already be divided by squaredLength.
+			var newX = (int)(lineEnd.X + (t * (lineStart.X - lineEnd.X) / squaredLength));
+			var newY = (int)(lineEnd.Y + (t * (lineStart.Y - lineEnd.Y) / squaredLength));
+			var newZ = (int)(lineEnd.Z + (t * (lineStart.Z - lineEnd.Z) / squaredLength));
+			var projection = new WPos(newX, newY, newZ);
```

Comment:
It doesn't need to, because at this point your longs are well within the int data range.  "(small number int) / (big number int) = 0." is the same for longs as it is for ints, and shouldn't be a problem here.